packages/firestore/src/core/sync_engine.ts

```diff
@@ -824,41 +832,80 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
+        activeTargets.toArray()
+      );
+      for (const queryData of activeQueries) {
+        this.remoteStore.listen(queryData);
+      }
     } else if (isPrimary === false && this.isPrimary !== false) {
       this.isPrimary = false;
-      await this.remoteStore.disableNetwork();
-      objUtils.forEachNumber(this.queryViewsByTarget, targetId => {
+      const activeQueries = await this.synchronizeQueryViewsAndRaiseSnapshots(
+        objUtils.indices(this.queryViewsByTarget)
+      );
+      this.resetLimboDocuments();
+      for (const queryData of activeQueries) {
         // TODO(multitab): Remove query views for non-local queries.
-        this.remoteStore.unlisten(targetId);
+        this.remoteStore.unlisten(queryData.targetId);
+      }
+      await this.remoteStore.applyPrimaryState(false);
+    }
+  }
+
+  // PORTING NOTE: Multi-tab only.
+  private resetLimboDocuments(): void {
+    objUtils.forEachNumber(this.limboKeysByTarget, targetId => {
+      this.remoteStore.unlisten(targetId);
+    });
+    this.limboKeysByTarget = [];
+    this.limboTargetsByKey = new SortedMap<DocumentKey, TargetId>(
+      DocumentKey.comparator
+    );
+  }
+
+  /**
+   * Reconcile the local query views with those the state from persistence.
+   * Raises snapshots for any changes that affect the current client.
+   */
+  // PORTING NOTE: Multi-tab only.
+  private synchronizeQueryViewsAndRaiseSnapshots(
+    targets: TargetId[]
+  ): Promise<QueryData[]> {
+    let p = Promise.resolve();
+    const activeQueries: QueryData[] = [];
+    for (const targetId of targets) {
+      p = p.then(async () => {
+        let queryData;
+        const query = await this.localStore.getQueryForTarget(targetId);
+        if (this.queryViewsByTarget[targetId] === undefined) {
+          assert(
+            this.isPrimary,
+            'A secondary tab should never have an active query without an active view.'
+          );
+          // For queries that never executed on this client, we need to
+          // allocate the query in LocalStore and initialize a new View.
+          queryData = await this.localStore.allocateQuery(query);
+          await this.initializeViewAndComputeSnapshot(queryData, false);
+        } else {
+          // For queries that have a local View, we need to update their state
+          // in LocalStore (as the resume token and the snapshot version
+          // might have changed) and reconcile their views with the persisted
+          // state (the list of syncedDocuments may have gotten out of sync).
+          await this.localStore.releaseQuery(query, true);
+          queryData = await this.localStore.allocateQuery(query);
+          const viewChange = await this.synchronizeViewAndComputeSnapshot(
+            query,
+            queryData.targetId
+          );
+          if (viewChange.snapshot) {
+            this.syncEngineListener!.onWatchChange([viewChange.snapshot]);
```

Comment:
Should we batch up all the snapshots and call onWatchChange() only once?

In practice I'm not sure how much it matters right now, but since all of the snapshots were the result of a single event, I think we should send them to onWatchChange() as a single batch.  In the future we might expose an API like:

```
firebase.firestore().onGlobalSnapshot(() => { /* I know that all listeners are in a consistent state. */ });
```